### DISCUSSION TITLE
gpt-4-0613 consistently exceeds maximum context length

### DISCUSSION THREAD

**Discussion by mark-asymbl** [danny-avila/LibreChat#1572] [Troubleshooting]
### What happened?

Every long conversation eventually results in:
> Something went wrong. Here's the specific error message we encountered: Error: 400 This model's maximum context length is 8192 tokens. However, your messages resulted in 8210 tokens. Please reduce the length of the messages.

The "However, your messages resulted in 8210 tokens" line varies in token count each time.

The commit I'm running: https://github.com/danny-avila/LibreChat/commit/406940490bd7c943781165a693cc7eeda8462f6d

### Steps to Reproduce

1. Start a new chat with gpt-4-0613 (I have not tested other models)
2. Continue the conversation for a while e.g. >30k tokens total including both your messages and the model's responses
3. Observe 
> Something went wrong. Here's the specific error message we encountered: Error: 400 This model's maximum context length is 8192 tokens. However, your messages resulted in <#> tokens. Please reduce the length of the messages.

### What browsers are you seeing the problem on?

Chrome

### Relevant log output

```shell
2024-01-17 01:12:23 warn: [OpenAIClient.chatCompletion][stream] API error
2024-01-17 01:12:23 warn: [OpenAIClient.chatCompletion][finalChatCompletion] API error
2024-01-17 01:12:23 error: [OpenAIClient.chatCompletion] Unhandled error type Error: 400 This model's maximum context length is 8192 tokens. However, your messages resulted in 8210 tokens. Please reduce the length of the messages.
2024-01-17 01:12:23 error: [handleAbortError] AI response error; aborting request: Error: 400 This model's maximum context length is 8192 tokens. However, your messages resulted in 8210 tokens. Please reduce the length of the messages.
```

### Screenshots

![Screenshot 2024-01-16 at 5 10 21 PM](https://github.com/danny-avila/LibreChat/assets/129498435/3812ba3a-5c48-4ac7-9a9d-719951df7321)
![Screenshot 2024-01-16 at 5 09 52 PM](https://github.com/danny-avila/LibreChat/assets/129498435/1e68d25f-1f6e-41d5-b8b7-8a8cfcefa09f)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
This is hard to reproduce at the >30k range. While I can certainly do it, it would be better with one of your exact examples. Are you comfortable sharing one of the conversations you already had? You can export via the settings menu which you access by clicking on your username. If you are, you can feel free to contact me privately to share the convo on one of my socials: https://github.com/danny-avila, preferably discord: https://discord.librechat.ai

It would also be helpful to see some debug logs from the `api/logs` and to inspect the `tokenCountMap` that is generated just before the completion fails.

It's not unexpected for the token count to be 1-2 tokens off, and at a range of 30k tokens, it makes sense that any miscounts would exceed at that level. I also noticed that `gpt-4-0613` only has a -1 buffer from the max, so that can definitely be adjusted to -5 or -10 to alleviate this.

**Reply by mark-asymbl** [reply]
> It's not unexpected for the token count to be 1-2 tokens off, and at a range of 30k tokens, it makes sense that any miscounts would exceed at that level. I also noticed that gpt-4-0613 only has a -1 buffer from the max, so that can definitely be adjusted to -5 or -10 to alleviate this.

The largest surplus I've seen so far is 18 tokens. However, while LibreChat has been a far superior experience so far otherwise, I've never had this problem with other chat interfaces that support the same model so I wonder if there's a more general, robust solution than a naive token buffer. Perhaps there's a lack of control due to dependence on third party libraries? I'll have to take a closer look at the implementation; I just deployed this for my team last night.

Short of a better solution, perhaps just catch the error, expand the buffer by the delta, and retry. Cache that buffer size locally to minimize wasted requests. Generate a warning log when this happens. Hard fail if it grows beyond a certain percentage of the model's supported context.

> This is hard to reproduce at the >30k range. While I can certainly do it, it would be better with one of your exact examples. Are you comfortable sharing one of the conversations you already had? You can export via the settings menu which you access by clicking on your username. If you are, you can feel free to contact me privately to share the convo on one of my socials: https://github.com/danny-avila, preferably discord: https://discord.librechat.ai/

Thanks for the offer! However, I'm unlikely to repro with anything not sensitive/proprietary anytime soon and I assume scrubbing an existing repro is impractical due to nondeterministic outputs. If I happen to, though, I'll send it. In the mean time, I'd rather invest the time fixing it locally, where I can repro, then submit a PR. Best if that's done based on your guidance before that time is spent though. Thoughts?

**Reply by danny-avila** [reply]
The problem is we can't get an accurate/official count when streaming is enabled, the API won't give us their amount. 

> The largest surplus I've seen so far is 18 tokens.

If it's off by 1 or 2 every couple of rounds, at 30k context that could easily be 10 miscalculations (off by one or 2) for X messages

> I wonder if there's a more general, robust solution than a naive token buffer

The handling of tokens matches this [OpenAI cookbook](https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb) exact, as can be seen here: https://github.com/danny-avila/LibreChat/blob/963dbf3a1e0cfdc84b81940f35ee1e3e1714bc47/api/app/clients/specs/OpenAIClient.test.js#L546

That said I don't know how much more robust we can get. The counting method is not wrong but maybe it's getting screwy somewhere along the pipeline.

Totally understand about the sensitive nature. At the very least, it will help me to see the `tokenCountMap` in your debug logs in `api/logs` directory, and if you were using a preset, using a custom name or instructions.

>  LibreChat has been a far superior experience so far otherwise

Thanks very much. This comes from a lot of the feedback along the way, and many helping hands. Glad it could serve you well!
